File: java-compat/src/Ice/src/main/java/IceInternal/DefaultsAndOverrides.java
// **********************************************************************
//
// Copyright (c) 2003-2017 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceInternal;

public final class DefaultsAndOverrides
{
    DefaultsAndOverrides(Ice.Properties properties, Ice.Logger logger)
    {
        String value;
        int intValue;

        defaultProtocol = properties.getPropertyWithDefault("Ice.Default.Protocol", "tcp");

        value = properties.getProperty("Ice.Default.Host");
        if(!value.isEmpty())
        {
            defaultHost = value;
        }
        else
        {
            defaultHost = null;
        }

        value = properties.getProperty("Ice.Default.SourceAddress");
        if(!value.isEmpty())
        {
            defaultSourceAddress = Network.getNumericAddress(value);
            if(defaultSourceAddress == null)
            {
                throw new Ice.InitializationException("invalid IP address set for Ice.Default.SourceAddress: `" +
                                                      value + "'");
            }
        }
        else
        {
            defaultSourceAddress = null;
        }

        value = properties.getProperty("Ice.Override.Timeout");
        if(!value.isEmpty())
        {
            overrideTimeout = true;
            intValue = properties.getPropertyAsInt("Ice.Override.Timeout");
            if(intValue < 0 && intValue != -1)
            {
                overrideTimeoutValue = -1;
                StringBuffer msg = new StringBuffer("invalid value for Ice.Override.Timeout `");
                msg.append(properties.getProperty("Ice.Override.Timeout"));
                msg.append("': defaulting to -1");
                logger.warning(msg.toString());
            }
            else
            {
               overrideTimeoutValue = intValue;
            }
        }
        else
        {
            overrideTimeout = false;
            overrideTimeoutValue = -1;
        }

        value = properties.getProperty("Ice.Override.ConnectTimeout");
        if(!value.isEmpty())
        {
            overrideConnectTimeout = true;
            intValue = properties.getPropertyAsInt("Ice.Override.ConnectTimeout");
            if(intValue < 0 && intValue != -1)
            {
                overrideConnectTimeoutValue = -1;
                StringBuffer msg = new StringBuffer("invalid value for Ice.Override.ConnectTimeout `");
                msg.append(properties.getProperty("Ice.Override.ConnectTimeout"));
                msg.append("': defaulting to -1");
                logger.warning(msg.toString());
            }
            else
            {
                overrideConnectTimeoutValue = intValue;
            }
        }
        else
        {
            overrideConnectTimeout = false;
            overrideConnectTimeoutValue = -1;
        }

        value = properties.getProperty("Ice.Override.CloseTimeout");
        if(!value.isEmpty())
        {
            overrideCloseTimeout = true;
            intValue = properties.getPropertyAsInt("Ice.Override.CloseTimeout");
            if(intValue < 0 && intValue != -1)
            {
                overrideCloseTimeoutValue = -1;
                StringBuffer msg = new StringBuffer("invalid value for Ice.Override.CloseTimeout `");
                msg.append(properties.getProperty("Ice.Override.CloseTimeout"));
                msg.append("': defaulting to -1");
                logger.warning(msg.toString());
            }
            else
            {
                overrideCloseTimeoutValue = intValue;
            }
        }
        else
        {
            overrideCloseTimeout = false;
            overrideCloseTimeoutValue = -1;
        }

        value = properties.getProperty("Ice.Override.Compress");
        if(!value.isEmpty())
        {
            overrideCompress = true;
            boolean b = properties.getPropertyAsInt("Ice.Override.Compress") > 0;
            if(b && !BZip2.supported())
            {
                System.err.println("warning: bzip2 support not available, Ice.Override.Compress ignored");
                b = false;
            }
            overrideCompressValue = b;
        }
        else
        {
            overrideCompress = false;
            overrideCompressValue = false;
        }

        value = properties.getProperty("Ice.Override.Secure");
        if(!value.isEmpty())
        {
            overrideSecure = true;
            overrideSecureValue = properties.getPropertyAsInt("Ice.Override.Secure") > 0;
        }
        else
        {
            overrideSecure = false;
            overrideSecureValue = false;
        }

        defaultCollocationOptimization =
            properties.getPropertyAsIntWithDefault("Ice.Default.CollocationOptimized", 1) > 0;

        value = properties.getPropertyWithDefault("Ice.Default.EndpointSelection", "Random");
        if(value.equals("Random"))
        {
            defaultEndpointSelection = Ice.EndpointSelectionType.Random;
        }
        else if(value.equals("Ordered"))
        {
            defaultEndpointSelection = Ice.EndpointSelectionType.Ordered;
        }
        else
        {
            Ice.EndpointSelectionTypeParseException ex = new Ice.EndpointSelectionTypeParseException();
            ex.str = "illegal value `" + value + "'; expected `Random' or `Ordered'";
            throw ex;
        }

        intValue = properties.getPropertyAsIntWithDefault("Ice.Default.Timeout", 60000);
        if(intValue < 1 && intValue != -1)
        {
            defaultTimeout = 60000;
            StringBuffer msg = new StringBuffer("invalid value for Ice.Default.Timeout `");
            msg.append(properties.getProperty("Ice.Default.Timeout"));
            msg.append("': defaulting to 60000");
            logger.warning(msg.toString());
        }
        else
        {
            defaultTimeout = intValue;
        }

        intValue = properties.getPropertyAsIntWithDefault("Ice.Default.LocatorCacheTimeout", -1);
        if(intValue < -1)
        {
            defaultLocatorCacheTimeout = -1;
            StringBuffer msg = new StringBuffer("invalid value for Ice.Default.LocatorCacheTimeout `");
            msg.append(properties.getProperty("Ice.Default.LocatorCacheTimeout"));
            msg.append("': defaulting to -1");
            logger.warning(msg.toString());
        }
        else
        {
            defaultLocatorCacheTimeout = intValue;
        }

        intValue = properties.getPropertyAsIntWithDefault("Ice.Default.InvocationTimeout", -1);
        if(intValue < 1 && intValue != -1 && intValue != -2)
        {
            defaultInvocationTimeout = -1;
            StringBuffer msg = new StringBuffer("invalid value for Ice.Default.InvocationTimeout `");
            msg.append(properties.getProperty("Ice.Default.InvocationTimeout"));
            msg.append("': defaulting to -1");
            logger.warning(msg.toString());
        }
        else
        {
            defaultInvocationTimeout = intValue;
        }

        defaultPreferSecure = properties.getPropertyAsIntWithDefault("Ice.Default.PreferSecure", 0) > 0;

        value = properties.getPropertyWithDefault("Ice.Default.EncodingVersion",
                                                  Ice.Util.encodingVersionToString(Protocol.currentEncoding));
        defaultEncoding = Ice.Util.stringToEncodingVersion(value);
        Protocol.checkSupportedEncoding(defaultEncoding);

        boolean slicedFormat = properties.getPropertyAsIntWithDefault("Ice.Default.SlicedFormat", 0) > 0;
        defaultFormat = slicedFormat ? Ice.FormatType.SlicedFormat : Ice.FormatType.CompactFormat;
    }

    final public String defaultHost;
    final public java.net.InetSocketAddress defaultSourceAddress;
    final public String defaultProtocol;
    final public boolean defaultCollocationOptimization;
    final public Ice.EndpointSelectionType defaultEndpointSelection;
    final public int defaultTimeout;
    final public int defaultLocatorCacheTimeout;
    final public int defaultInvocationTimeout;
    final public boolean defaultPreferSecure;
    final public Ice.EncodingVersion defaultEncoding;
    final public Ice.FormatType defaultFormat;

    final public boolean overrideTimeout;
    final public int overrideTimeoutValue;
    final public boolean overrideConnectTimeout;
    final public int overrideConnectTimeoutValue;
    final public boolean overrideCloseTimeout;
    final public int overrideCloseTimeoutValue;
    final public boolean overrideCompress;
    final public boolean overrideCompressValue;
    final public boolean overrideSecure;
    final public boolean overrideSecureValue;
}

File: csharp/src/Ice/DefaultsAndOverrides.cs
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

using Ice;
using System.Net;

namespace IceInternal
{
    public sealed class DefaultsAndOverrides
    {
        internal DefaultsAndOverrides(Communicator communicator, ILogger logger)
        {
            string? val;

            DefaultTransport = communicator.GetProperty("Ice.Default.Transport") ?? "tcp";

            DefaultHost = communicator.GetProperty("Ice.Default.Host");

            val = communicator.GetProperty("Ice.Default.SourceAddress");
            if (val != null)
            {
                DefaultSourceAddress = Network.GetNumericAddress(val);
                if (DefaultSourceAddress == null)
                {
                    throw new InvalidConfigurationException(
                        $"invalid IP address set for Ice.Default.SourceAddress: `{val}'");
                }
            }
            else
            {
                DefaultSourceAddress = null;
            }

            val = communicator.GetProperty("Ice.Override.Timeout");
            if (val != null)
            {
                OverrideTimeout = true;
                OverrideTimeoutValue = communicator.GetPropertyAsInt("Ice.Override.Timeout") ?? 0;
                if (OverrideTimeoutValue < 1 && OverrideTimeoutValue != -1)
                {
                    OverrideTimeoutValue = -1;
                    var msg = new System.Text.StringBuilder("invalid value for Ice.Override.Timeout `");
                    msg.Append(communicator.GetProperty("Ice.Override.Timeout"));
                    msg.Append("': defaulting to -1");
                    logger.Warning(msg.ToString());
                }
            }
            else
            {
                OverrideTimeout = false;
                OverrideTimeoutValue = -1;
            }

            val = communicator.GetProperty("Ice.Override.ConnectTimeout");
            if (val != null)
            {
                OverrideConnectTimeout = true;
                OverrideConnectTimeoutValue = communicator.GetPropertyAsInt("Ice.Override.ConnectTimeout") ?? -1;
                if (OverrideConnectTimeoutValue < 1 && OverrideConnectTimeoutValue != -1)
                {
                    OverrideConnectTimeoutValue = -1;
                    var msg = new System.Text.StringBuilder("invalid value for Ice.Override.ConnectTimeout `");
                    msg.Append(communicator.GetProperty("Ice.Override.ConnectTimeout"));
                    msg.Append("': defaulting to -1");
                    logger.Warning(msg.ToString());
                }
            }
            else
            {
                OverrideConnectTimeout = false;
                OverrideConnectTimeoutValue = -1;
            }

            val = communicator.GetProperty("Ice.Override.CloseTimeout");
            if (val != null)
            {
                OverrideCloseTimeout = true;
                OverrideCloseTimeoutValue = communicator.GetPropertyAsInt("Ice.Override.CloseTimeout") ?? -1;
                if (OverrideCloseTimeoutValue < 1 && OverrideCloseTimeoutValue != -1)
                {
                    OverrideCloseTimeoutValue = -1;
                    var msg = new System.Text.StringBuilder("invalid value for Ice.Override.CloseTimeout `");
                    msg.Append(communicator.GetProperty("Ice.Override.CloseTimeout"));
                    msg.Append("': defaulting to -1");
                    logger.Warning(msg.ToString());
                }
            }
            else
            {
                OverrideCloseTimeout = false;
                OverrideCloseTimeoutValue = -1;
            }

            val = communicator.GetProperty("Ice.Override.Compress");
            if (val != null)
            {
                OverrideCompress = true;
                OverrideCompressValue = communicator.GetPropertyAsInt("Ice.Override.Compress") > 0;
                if (!BZip2.IsLoaded && OverrideCompressValue)
                {
                    logger.Warning("compression not supported bzip2 library not found, Ice.Override.Compress ignored");
                    OverrideCompressValue = false;
                }
            }
            else
            {
                OverrideCompress = !BZip2.IsLoaded;
                OverrideCompressValue = false;
            }

            DefaultCollocationOptimization = (communicator.GetPropertyAsInt("Ice.Default.CollocationOptimized") ?? 1) > 0;

            val = communicator.GetProperty("Ice.Default.EndpointSelection") ?? "Random";
            if (val.Equals("Random"))
            {
                DefaultEndpointSelection = Ice.EndpointSelectionType.Random;
            }
            else if (val.Equals("Ordered"))
            {
                DefaultEndpointSelection = EndpointSelectionType.Ordered;
            }
            else
            {
                throw new InvalidConfigurationException($"illegal value `{val}'; expected `Random' or `Ordered'");
            }

            DefaultTimeout = communicator.GetPropertyAsInt("Ice.Default.Timeout") ?? 60000;
            if (DefaultTimeout < 1 && DefaultTimeout != -1)
            {
                DefaultTimeout = 60000;
                var msg = new System.Text.StringBuilder("invalid value for Ice.Default.Timeout `");
                msg.Append(communicator.GetProperty("Ice.Default.Timeout"));
                msg.Append("': defaulting to 60000");
                logger.Warning(msg.ToString());
            }

            DefaultLocatorCacheTimeout = communicator.GetPropertyAsInt("Ice.Default.LocatorCacheTimeout") ?? -1;
            if (DefaultLocatorCacheTimeout < -1)
            {
                DefaultLocatorCacheTimeout = -1;
                var msg = new System.Text.StringBuilder("invalid value for Ice.Default.LocatorCacheTimeout `");
                msg.Append(communicator.GetProperty("Ice.Default.LocatorCacheTimeout"));
                msg.Append("': defaulting to -1");
                logger.Warning(msg.ToString());
            }

            DefaultInvocationTimeout = communicator.GetPropertyAsInt("Ice.Default.InvocationTimeout") ?? -1;
            if (DefaultInvocationTimeout < 1 && DefaultInvocationTimeout != -1)
            {
                DefaultInvocationTimeout = -1;
                var msg = new System.Text.StringBuilder("invalid value for Ice.Default.InvocationTimeout `");
                msg.Append(communicator.GetProperty("Ice.Default.InvocationTimeout"));
                msg.Append("': defaulting to -1");
                logger.Warning(msg.ToString());
            }

            // TODO: switch to 0 default
            DefaultPreferNonSecure = (communicator.GetPropertyAsInt("Ice.Default.PreferNonSecure") ?? 1) > 0;

            val = communicator.GetProperty("Ice.Default.Encoding");
            if (val == null)
            {
                DefaultEncoding = Encoding.Latest;
            }
            else
            {
                DefaultEncoding = Encoding.Parse(val);
                DefaultEncoding.CheckSupported();
            }

            bool slicedFormat = communicator.GetPropertyAsInt("Ice.Default.SlicedFormat") > 0;
            DefaultFormat = slicedFormat ? Ice.FormatType.Sliced : Ice.FormatType.Compact;
        }

        public string? DefaultHost;
        public IPAddress? DefaultSourceAddress;
        public string DefaultTransport;
        public bool DefaultCollocationOptimization;
        public EndpointSelectionType DefaultEndpointSelection;
        public int DefaultTimeout;
        public int DefaultLocatorCacheTimeout;
        public int DefaultInvocationTimeout;
        public bool DefaultPreferNonSecure;
        public Encoding DefaultEncoding;
        public FormatType DefaultFormat;

        public bool OverrideTimeout;
        public int OverrideTimeoutValue;
        public bool OverrideConnectTimeout;
        public int OverrideConnectTimeoutValue;
        public bool OverrideCloseTimeout;
        public int OverrideCloseTimeoutValue;
        public bool OverrideCompress;
        public bool OverrideCompressValue;
    }

}
